x-pack/legacy/plugins/epm/server/datasources/create.ts

```diff
@@ -38,30 +38,61 @@ export async function saveDatasourceReferences(options: {
 }) {
   const { savedObjectsClient, pkgkey, toSave } = options;
   const savedObject = await getDatasourceObject({ savedObjectsClient, pkgkey });
-  const savedRefs = savedObject && savedObject.attributes.installed;
-  const mergeRefsReducer = (current: AssetReference[], pending: AssetReference) => {
+  const savedRefs = savedObject?.attributes.package.assets;
+  const mergeRefsReducer = (current: Asset[] = [], pending: Asset) => {
     const hasRef = current.find(c => c.id === pending.id && c.type === pending.type);
     if (!hasRef) current.push(pending);
     return current;
   };
 
-  const toInstall = toSave.reduce(mergeRefsReducer, savedRefs || []);
-
-  await savedObjectsClient.create<DatasourceAttributes>(
-    SAVED_OBJECT_TYPE_DATASOURCES,
-    { installed: toInstall },
-    { id: pkgkey, overwrite: true }
-  );
+  const toInstall = (toSave as Asset[]).reduce(mergeRefsReducer, savedRefs);
+  const datasource: Datasource = createFakeDatasource(pkgkey, toInstall);
+  await savedObjectsClient.create<Datasource>(SAVED_OBJECT_TYPE_DATASOURCES, datasource, {
+    id: pkgkey,
+    overwrite: true,
+  });
 
   return toInstall;
 }
 
 export async function getDatasourceObject(options: {
   savedObjectsClient: SavedObjectsClientContract;
   pkgkey: string;
-}): Promise<DatasourceSavedObject | undefined> {
+}) {
   const { savedObjectsClient, pkgkey } = options;
   return savedObjectsClient
```

Comment:
I left this as-is to keep the changes as small as possible, but it occurs to me that we could also do

```typescript
const savedObject = await savedObjectsClient
  .get<Datasource>(SAVED_OBJECT_TYPE_DATASOURCES, pkgkey)

return savedObject.attributes
```

and return the `Datasource` type/object instead of the Saved Object and it's additional outer properties (`error?`, `migrationVersion`, etc). 

At the moment, there's only one consumer of this function and it's not interested in the Saved Object; only the `Datasource`. It'd be nice if it didn't have to "know" about a Saved Object shape.

Reply:
Agree with keeping it as is in this PR but follow up later and directly return the datasource. At the moment can't see a reason why we would be interested in the details of the outer properties.